src/asyncrpcoperation.h

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2016 The Zcash developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+
+#ifndef ASYNCRPCOPERATION_H
+#define ASYNCRPCOPERATION_H
+
+#include <string>
+#include <atomic>
+#include <map>
+#include <chrono>
+
+#include "json/json_spirit_value.h"
+#include "json/json_spirit_utils.h"
+#include "json/json_spirit_reader_template.h"
+#include "json/json_spirit_writer_template.h"
+
+using namespace std;
+using namespace json_spirit;
+
+/**
+ * AsyncRPCOperation objects are submitted to the AsyncRPCQueue for processing.
+ * 
+ * To subclass AsyncRPCOperation, implement the main() method.
+ * Update the operation status as work is underway and completes.
+ * If main() can be interrupted, inmplement the cancel() method.
+ */
+
+typedef std::string AsyncRPCOperationId;
+
+typedef enum class operationStateEnum {
+    READY = 0,
+    EXECUTING,
+    CANCELLED,
+    FAILED,
+    SUCCESS
+} OperationStatus;
+
+class AsyncRPCOperation {
+public:
+    AsyncRPCOperation();
+    virtual ~AsyncRPCOperation();
+
+    // You must implement this method in your subclass.
+    virtual void main();
+
+    // Override this method if you can interrupt execution of main() in your subclass.
+    void cancel();
+    
+    // Getters and setters
+
+    OperationStatus getState() const {
+        return state_.load();
+    }
+        
+    AsyncRPCOperationId getId() const {
+        return id_;
+    }
+   
+    int64_t getCreationTime() const {
+        return creation_time_;
+    }
+
+    Value getStatus() const;
+
+    Value getError() const;
+    
+    Value getResult() const;
+
+    std::string getStateAsString() const;
+    
+    int getErrorCode() const {
+        return error_code_;
+    }
+
+    std::string getErrorMessage() const {
+        return error_message_;
+    }
+
+    bool isCancelled() const {
+        return OperationStatus::CANCELLED==getState();
+    }
+
+    bool isExecuting() const {
+        return OperationStatus::EXECUTING==getState();
+    }
+
+    bool isReady() const {
+        return OperationStatus::READY==getState();
+    }
+
+    bool isFailed() const {
+        return OperationStatus::FAILED==getState();
+    }
+    
+    bool isSuccess() const {
+        return OperationStatus::SUCCESS==getState();
+    }
+
+protected:
+
+    Value result_;
+    int error_code_;
+    std::string error_message_;
+    std::atomic<OperationStatus> state_;
```

Comment:
Is state_ the only member that needs to be accessed atomically? I would expect that a lock is needed: remember that even if you observe state_ to have changed and know that main() changed another of the variables prior to that, you are *not* guaranteed to correctly see the change to that other variable unless it is itself atomic, or the accesses are locked.